### PR TITLE
[WIP] Upload progress bar now displays informations about uploads...

### PR DESCRIPTION
**Pull request by luckydonald**
Upload progress bar now displays informations about uploads, uploades size (e.g. 12kB of 3MB) and the current speed.

Making this PR for this: #8105 

Here is a live test:
user: test
pw: bla
http://flutterb.at/owncloud-testserver

should replace https://github.com/owncloud/core/pull/10324

### CODE REVIEW

**Comment by PVince81** [comment]
Here is the doc how to run JS unit tests locally: http://doc.owncloud.org/server/7.0/developer_manual/core/unit-testing.html#javascript-unit-testing-for-core

The only unit tests for upload are here: https://github.com/owncloud/core/blob/master/apps/files/tests/js/fileUploadSpec.js
But it might be a bit tricky to implement if you're not familiar with writing JS unit tests.

The idea would be to simulate a download start event (like done in the existing tests), then simulate the "fileuploadprogress" event on the `$dummyUploader` object passing some sample progress data, then verifying that the UI has been updated accordingly. Also you'd need to use `sinon.useFakeTimers()` to simulate time passing.

**Comment by PVince81** [comment]
Please rebase.
OC 8 is out, so this is a chance to get this in 8.1!

Any luck with the unit tests ?
I suggest you move the progress bar rendering logic into a separate function and then write a unit test for that function separately. Might be easier to do than simulating the "fileuploadprogressall" event.

**Comment by enoch85** [comment]
ping @luckydonald

**Comment by luckydonald** [comment]
hmm, Will get some time to look into this the next month or so.

**Comment by luckydonald** [comment]
Well,
I am actually avoiding grappling with the big topic 'unit tests' at the moment...

**Comment by ts78** [comment]
Was just searching for exactly that feature because I am really missing some indicators in Owncloud when uploading files - I don't see how many files are uploaded, how long it will take, how fast my internet connection is (or if I have a network problem) and so on. I like how it is implemented at mega.nz for example.

So it would be great if someone could support luckydonald with this to get it in next releases!

**Comment by luckydonald** [comment]
I am too afraid to touch that test stuff, basically because I have no idea what to do.
Maybe this could be done by someone having more experience with that.
Thats why there is no progress.

File: apps/files/js/file-upload.js [hunk 1]
```diff
@@ -464,11 +481,67 @@ OC.Upload = {
...
+						buffer[bufferIndex] = remainingSeconds; //buffer to make it smoother
+						bufferIndex = (bufferIndex + 1) % bufferSize;
+					}
+					var smoothRemainingSeconds = (bufferTotal / bufferSize); //seconds
+					var date = new Date(smoothRemainingSeconds * 1000);
+					var timeStringDesktop = "";
+					var timeStringMobile = ""; 
+					if(date.getUTCHours() > 0){
+						timeStringDesktop = t('files', '{hours}:{minutes}:{seconds} hour{plural_s} left' , { 
+							hours:date.getUTCHours(),
```
Comment (PVince81) [inline apps/files/js/file-upload.js]: I think we can now use momentjs for that which is included in core, see http://momentjs.com/

**Comment by PVince81** [comment]
@luckydonald needs rebasing due to conflicts. Let me know if you need help.

Also, if I can provide you with a skeleton for the unit tests, would you be able to write them ? (it looks like this https://github.com/owncloud/core/blob/master/apps/files/tests/js/fileUploadSpec.js#L95

I can take care of the hardest part which is setting up the base code for the tests to simulate upload until it reaches your function (the one that have several code paths that needs testing).

**Comment by PVince81** [comment]
another idea would be to move the code that does the complex calculation into a separate function OC.Upload.calculateTime(...). Then the unit test could be written for only testing that function.

**Comment by enoch85** [comment]
@PVince81 Yes, please help (if you ask me). Would really like this in 8.2!

**Comment by luckydonald** [comment]
@Pvince yeah, that would be probably better to separate the calculation.

I copied the calculation part from the desktop application, without in-depth understanding of the math behind that.

**Comment by DeepDiver1975** [comment]
@PVince81 I guess because of the latest development a rebase is out of question - right?

Can we revive these changes somehow?

@luckydonald

**Comment by luckydonald** [comment]
The problem where is hanging is I have no clue on how to write tests.

**Comment by PVince81** [comment]
I was hoping to maybe integrate this later with the Webdav upload: https://github.com/owncloud/core/pull/21237

**Comment by PVince81** [comment]
Later as in 9.1 because I don't have time to finish webdav upload unfortunately.
That branch still uses blueimp so the code should be reusable.

Also my branch cleans up the upload code which makes it easier to write tests.

**Comment by karlitschek** [comment]
@luckydonald We can merge this without unit tests. Can you please send a contributor agreement or state here in github that that this is MIT licensed?
@rullzer Can you help with rebase?

**Comment by luckydonald** [comment]
I think I already stated that before via Email, but here you go again:
I confirm the MIT for this contribution.

**Comment by PVince81** [comment]
Here we go, I rebased this for you and resolved the conflicts.
Moved to a separate PR for CI: https://github.com/owncloud/core/pull/24605

Please go there and review it, then we can merge this for 9.1.
Thanks!

**Comment by MorrisJobke** [comment]
Thanks a lot @luckydonald :) Sorry, that this took so long :(

**Comment by luckydonald** [comment]
I couldn't get around to help here, too, so thats no problem 😄

**Pull request closed without merging**